### PR TITLE
8331907: BigInteger and BigDecimal should use optimized division

### PR DESCRIPTION
**Pull request by djelinski**
Replace the custom unsigned divide / remainder functions with the compiler-optimized Long.divideUnsigned / remainderUnsigned.

No new tests. Existing tier1-3 tests continue to pass.

I added a new set of divide benchmarks. Results in thread.

I also removed the BigDecimal.toString benchmarks. They no longer serve their purpose - toString caches the returned value, so we were only benchmarking the cache access time.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8331907](https://bugs.openjdk.org/browse/JDK-8331907): BigInteger and BigDecimal should use optimized division (**Enhancement** - P4)

### Reviewers
 * [Raffaello Giulietti](https://openjdk.org/census#rgiulietti) (@rgiulietti - **Reviewer**)
 * [Brian Burkhalter](https://openjdk.org/census#bpb) (@bplb - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/19134/head:pull/19134` \
`$ git checkout pull/19134`

Update a local copy of the PR: \
`$ git checkout pull/19134` \
`$ git pull https://git.openjdk.org/jdk.git pull/19134/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 19134`

View PR using the GUI difftool: \
`$ git pr show -t 19134`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/19134.diff">https://git.openjdk.org/jdk/pull/19134.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/19134#issuecomment-2100047222)

### CODE REVIEW

**Comment by djelinski** [comment]
Results before:
```
Benchmark                                   Mode  Cnt    Score   Error  Units
BigDecimals.testHugeLargeDivide             avgt   15  115.176 ± 0.965  ns/op
BigDecimals.testHugeSmallDivide             avgt   15   82.652 ± 0.601  ns/op
BigDecimals.testLargeSmallDivide            avgt   15    6.601 ± 0.320  ns/op
BigIntegers.testHugeLargeDivide             avgt   15   53.905 ± 0.734  ns/op
BigIntegers.testHugeSmallDivide             avgt   15   52.762 ± 0.354  ns/op
BigIntegers.testLargeSmallDivide            avgt   15   22.990 ± 0.058  ns/op
```
after:
```
Benchmark                                   Mode  Cnt    Score   Error  Units
BigDecimals.testHugeLargeDivide             avgt   15  106.549 ± 0.695  ns/op
BigDecimals.testHugeSmallDivide             avgt   15   68.339 ± 0.172  ns/op
BigDecimals.testLargeSmallDivide            avgt   15    6.594 ± 0.328  ns/op
BigIntegers.testHugeLargeDivide             avgt   15   29.576 ± 0.411  ns/op
BigIntegers.testHugeSmallDivide             avgt   15   30.072 ± 0.242  ns/op
BigIntegers.testLargeSmallDivide            avgt   15    8.034 ± 0.078  ns/op
```

File: src/java.base/share/classes/java/math/BigDecimal.java [hunk 1]
```diff
@@ -5681,17 +5681,8 @@ private static BigDecimal divideAndRound128(final long dividendHi, final long di
         tmp = (dividendHi << shift) | (dividendLo >>> 64 - shift);
         long u2 = tmp & LONG_MASK;
         long q1, r_tmp;
```
Comment (TheShermanTanker) [inline src/java.base/share/classes/java/math/BigDecimal.java:5683]: Nit: There doesn't seem to be a reason for splitting the q1 and r\_tmp declarations from their assignments, why not the following instead? (Also applies to the cases below)
```
long q1 = Long.divideUnsigned(tmp, v1);
long r_tmp = Long.remainderUnsigned(tmp, v1);
```

File: src/java.base/share/classes/java/math/MutableBigInteger.java [hunk 2]
```diff
@@ -1106,41 +1106,19 @@ int divideOneWord(int divisor, MutableBigInteger quotient) {
...
-                rem = (int) (dividendEstimate - q * divisorLong);
-            } else {
-                long tmp = divWord(dividendEstimate, divisor);
-                q = (int) (tmp & LONG_MASK);
-                rem = (int) (tmp >>> 32);
-            }
+            int q = (int) Long.divideUnsigned(dividendEstimate, divisorLong);
+            rem = (int) Long.remainderUnsigned(dividendEstimate, divisorLong);
             quotient.value[intLen - xlen] = q;
             remLong = rem & LONG_MASK;
```
Comment (rgiulietti) [inline src/java.base/share/classes/java/math/MutableBigInteger.java:1117]: What about removing one of the `int rem`, renaming `remLong` to `rem` and using it as follows?

```
        long rem = 0;
        for (int xlen = intLen; xlen > 0; xlen--) {
            long dividendEstimate = (rem << 32) |
                    (value[offset + intLen - xlen] & LONG_MASK);
            int q = (int) Long.divideUnsigned(dividendEstimate, divisorLong);
            rem = Long.remainderUnsigned(dividendEstimate, divisorLong);
            quotient.value[intLen - xlen] = q;
        }

        quotient.normalize();
        return (int) rem;
```

**Comment by rgiulietti** [comment]
There's another spot for replacement at L.1096-1097 in `MutableBigInteger`. Did you see that one?

**Comment by djelinski** [comment]
L.1096-1097 is a 32-bit by 32-bit division; I guess I could change it to use Integer.unsignedDivide. Let me look into it...

**Review by rgiulietti — APPROVED** [review]
LGTM

**Review by bplb — APPROVED** [review]
I have nothing to add to the comments already resolved. +1

**Comment by djelinski** [comment]
Thanks for the reviews!

/integrate

**Pull request closed without merging**